File: design_patterns_101/Structural Patterns/Adapter Pattern/adapterPattern.js
// Target interface
class Target {
    request() {
      return "Target: The default target's behavior.";
    }
  }
  
  // Adaptee
  class Adaptee {
    specificRequest() {
      return ".eetpadA eht fo roivaheb laicepS";
    }
  }
  
  // Adapter
  class Adapter extends Target {
    constructor(adaptee) {
      super();
      this.adaptee = adaptee;
    }
  
    request() {
      const result = this.adaptee.specificRequest().split("").reverse().join("");
      return `Adapter: (TRANSLATED) ${result}`;
    }
  }
  
  // Usage
  function clientCode(target) {
    console.log(target.request());
  }
  
  const adaptee = new Adaptee();
  const adapter = new Adapter(adaptee);
  
  console.log("Adaptee interface is incompatible with the client.");
  console.log(`But with Adapter ${adapter.request()}`);
  
  clientCode(adapter);
  

File: design_patterns_101/Structural Patterns/Adapter Pattern/adapterPattern.py
class Target:
    def request(self) -> str:
        return "Target: The default target's behavior."

class Adaptee:
    def specific_request(self) -> str:
        return ".eetpadA eht fo roivaheb laicepS"

class Adapter(Target, Adaptee):
    def request(self) -> str:
        return f"Adapter: (TRANSLATED) {self.specific_request()[::-1]}"

# Usage
def client_code(target: Target) -> None:
    print(target.request())

adaptee = Adaptee()
adapter = Adapter()

print("Adaptee interface is incompatible with the client.")
print(f"But with Adapter {adapter.request()}")

client_code(adapter)
